paddle/fluid/framework/ir/memory_optimize_pass/buffer_shared_identity_inplace_op_pass.cc

```diff
@@ -0,0 +1,354 @@
+// Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <queue>
+#include "paddle/fluid/framework/ir/memory_optimize_pass/memory_reuse_pass.h"
+
+namespace paddle {
+namespace framework {
+namespace ir {
+
+using OpHandleBase = details::OpHandleBase;
+using ComputationOpHandle = details::ComputationOpHandle;
+using VarHandle = details::VarHandle;
+using VarHandleBase = details::VarHandleBase;
+
+/**
+ * The vars in `IdentityInplaceVarInfo` builds a `FOREST`.
+ *
+ * This means that:
+ *  1. there is no circle. Otherwise, it must be a bug of Paddle framework.
+ *  2. input edge number of any var must be 0 (for root) or 1.
+ *  3. output edge number of any var must be 0 (for leaf) or larger than 0.
+ *
+ * We would use the characteristics of `FOREST` in the following codes,
+ * so we prove these 3 conditions here.
+ *
+ * Condition 1 and 3 is obvious. Let us prove condition 2:
+ *
+ *  In order to perform in-place reuse, `output_var` must be the 1st version
+ *  in SSA graph (it is guaranteed in `MemoryReusePass::IsOutVarReusable`
+ *  method). Therefore, if the input edge number of `output_var` is more than 1,
+ *  it means `output_var` must be outputs of at least 2 different ops (one op
+ *  with the same output is not allowed in Paddle framework). However, if
+ *  `output_var` is the output of 2 different ops, `output_var`s in these 2 ops
+ *  must have different version, which is guaranteed by SSA graph. Therefore,
+ *  `output_var` must be the unique output of unique op. Condition 2 is proved.
+ */
+using IdentityInplaceVarInfo = std::unordered_map<
+    VarHandle * /*input_var*/,
+    std::vector<std::pair<VarHandle * /*output_var*/,
+                          ComputationOpHandle * /*identity op*/>>>;
+
+static std::unordered_map<VarHandle *, VarHandle *> BuildParent(
+    const IdentityInplaceVarInfo &info) {
+  std::unordered_map<VarHandle *, VarHandle *> parent;
+
+  for (auto &pair : info) {
+    PADDLE_ENFORCE_EQ(pair.second.empty(), false,
+                      platform::errors::InvalidArgument(
+                          "Var info list cannot be empty, this may be a bug"));
+
+    auto *in_var = pair.first;
+    parent[in_var];
```

Comment:
It is put "in_var" as key in parent if it is not in parent

Reply:
I got it, while why we need this, i.e. insert a `key` without `value`?

Reply:
Yes, you are right. This line is used to adding `nullptr` parent for root vars. It is useful because we can distinguish whether the `var` is inside the forest or not to avoid probable bug.